influxdb3_write/src/last_cache.rs

```diff
@@ -0,0 +1,498 @@
+    CacheAlreadyExists,
+}
+
+/// A two level hashmap storing Database Name -> Table Name -> LastCache
+///
+/// There are two lock levels, one at the top and one at the bottom:
+/// - Top: lock the entire cache for creating new entries
+/// - Bottom: lock an individual cache for pushing in new data
+type CacheMap = RwLock<HashMap<String, HashMap<String, RwLock<LastCache>>>>;
+
+pub struct LastCacheProvider {
+    cache_map: CacheMap,
+}
+
+impl std::fmt::Debug for LastCacheProvider {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "LastCacheProvider")
+    }
+}
+
+impl LastCacheProvider {
+    /// Create a new [`LastCacheProvider`]
+    pub(crate) fn new() -> Self {
+        Self {
+            cache_map: Default::default(),
+        }
+    }
+
+    /// Output the records for a given cache as arrow [`RecordBatch`]es
+    #[cfg(test)]
+    pub(crate) fn get_cache_record_batches<D, T>(
+        &self,
+        db_name: D,
+        tbl_name: T,
+    ) -> Option<Result<RecordBatch, ArrowError>>
+    where
+        D: AsRef<str>,
+        T: AsRef<str>,
+    {
+        self.cache_map
+            .read()
+            .get(db_name.as_ref())
+            .and_then(|db| db.get(tbl_name.as_ref()))
+            .map(|lc| lc.read().to_record_batch())
+    }
+
+    /// Create a new entry in the last cache for a given database and table, along with the given
+    /// parameters.
+    pub(crate) fn create_cache<D, T>(
+        &self,
+        db_name: D,
+        tbl_name: T,
+        count: usize,
+        key_columns: impl IntoIterator<Item: Into<String>>,
+        schema: SchemaRef,
+    ) -> Result<(), Error>
+    where
+        D: Into<String>,
+        T: Into<String>,
+    {
+        let db_name = db_name.into();
+        let tbl_name = tbl_name.into();
+        if self
+            .cache_map
+            .read()
+            .get(&db_name)
+            .is_some_and(|db| db.contains_key(&tbl_name))
+        {
+            return Err(Error::CacheAlreadyExists);
+        }
+        let last_cache = RwLock::new(LastCache::new(count, key_columns, schema)?);
+        self.cache_map
+            .write()
+            .entry(db_name)
+            .or_default()
+            .insert(tbl_name, last_cache);
+        Ok(())
+    }
+
+    /// Write a batch from the buffer into the cache by iterating over its database and table batches
+    /// to find entries that belong in the cache.
+    ///
+    /// Only if rows are newer than the latest entry in the cache will they be entered.
+    pub(crate) fn write_batch_to_cache(&self, write_batch: &WriteBatch) {
+        for (db_name, db_batch) in &write_batch.database_batches {
+            for (tbl_name, tbl_batch) in &db_batch.table_batches {
+                if let Some(db) = self.cache_map.read().get(db_name.as_str()) {
+                    for row in &tbl_batch.rows {
+                        if db
+                            .get(tbl_name)
```

Comment:
better to get the table above the loop since multiple rows could be in the table batch.